pkg/cache/cache.go

```diff
@@ -17,333 +17,110 @@ limitations under the License.
-	}
-	return nil
+	// IndexField adds an index with the given field name on the given object type
+	// by using the given function to extract the value for that field.  If you want
+	// compatibility with the Kubernetes API server, only return one key, and only use
+	// fields that the API server supports.  Otherwise, you can return multiple keys,
+	// and "equality" in the field selector means that at least one key matches the value.
+	IndexField(obj runtime.Object, field string, extractValue client.IndexerFunc) error
 }
 
-func (o *objectCache) cacheFor(obj runtime.Object) (*singleObjectCache, error) {
-	if !o.informers.started {
-		return nil, fmt.Errorf("must start Cache before calling Get or List %s %s",
-			"Object", fmt.Sprintf("%T", obj))
-	}
-	objType := reflect.TypeOf(obj)
+// Options are the optional arguments for creating a new InformersMap object
+type Options struct {
+	// Scheme is the scheme to use for mapping objects to GroupVersionKinds
+	Scheme *runtime.Scheme
 
-	cache, isKnown := o.cachesByType[objType]
-	if !isKnown {
-		return nil, fmt.Errorf("no Cache found for %T - must call GetInformer", obj)
-	}
-	return cache, nil
-}
+	// Mapper is the RESTMapper to use for mapping GroupVersionKinds to Resources
+	Mapper meta.RESTMapper
 
-// Get implements populatingClient.Reader
-func (o *objectCache) Get(ctx context.Context, key client.ObjectKey, out runtime.Object) error {
-	// Make sure there is a Cache for this type
-	if !o.has(out) {
-		err := o.init(out)
-		if err != nil {
-			return err
-		}
-	}
-
-	cache, err := o.cacheFor(out)
-	if err != nil {
-		return err
-	}
-
-	return cache.Get(ctx, key, out)
-}
-
-// List implements populatingClient.Reader
-func (o *objectCache) List(ctx context.Context, opts *client.ListOptions, out runtime.Object) error {
-	itemsPtr, err := apimeta.GetItemsPtr(out)
-	if err != nil {
-		return nil
-	}
-
-	ro, ok := itemsPtr.(runtime.Object)
-	if ok && !o.has(ro) {
-		err = o.init(ro)
-		if err != nil {
-			return err
-		}
-	}
-
-	// http://knowyourmeme.com/memes/this-is-fine
-	outType := reflect.Indirect(reflect.ValueOf(itemsPtr)).Type().Elem()
-	cache, isKnown := o.cachesByType[outType]
-	if !isKnown {
-		return fmt.Errorf("no cache for objects of type %T", out)
-	}
-	return cache.List(ctx, opts, out)
+	// Resync is the resync period
+	Resync *time.Duration
 }
 
-// singleObjectCache is a Reader
-var _ client.Reader = &singleObjectCache{}
+var _ Informers = &informerCache{}
+var _ client.Reader = &informerCache{}
+var _ Cache = &informerCache{}
 
-// singleObjectCache is a Reader that retrieves objects
-// from a single local cache populated by a watch.
-type singleObjectCache struct {
-	// Indexer is the underlying indexer wrapped by this cache.
-	Indexer cache.Indexer
-	// GroupVersionKind is the group-version-kind of the resource.
-	GroupVersionKind schema.GroupVersionKind
+// cache is a Kubernetes Object cache populated from InformersMap.  cache wraps a CacheProvider and InformersMap.
```

Comment:
It looks like informerCache only wraps an InformersMap, not a CacheProvider as stated in comment.

Reply:
Done